File: PHP/StdLib/PackagesFolderSource.php
<?php

namespace StdLib\PackagesFolderSource;

use onepkg\OneFile\OneFile;
use StdLib\PackageManager\PackageSource;
use StdLib\PackageManager\PackageId;
use StdLib\PackageManager\PackageBundle;
use StdLib\PackageManager\PackageContent;
use StdLib\PackageManager\PackageType;

class PackagesFolderSource implements PackageSource {
    public $packagesDir;
    
    function __construct($packagesDir = "packages") {
        $this->packagesDir = $packagesDir;
    }
    
    function getPackageBundle($ids, $cachedOnly) {
        throw new \OneLang\Error("Method not implemented.");
    }
    
    function getAllCached() {
        $packages = Array();
        $allFiles = OneFile::listFiles($this->packagesDir, true);
        foreach ($allFiles as $fn) {
            if ($fn === "bundle.json")
                continue;
            // TODO: hack
            $pathParts = preg_split("/\\//", $fn);
            // [0]=implementations/interfaces, [1]=package-name, [2:]=path
            $type = array_shift($pathParts);
            $pkgDir = array_shift($pathParts);
            if ($type !== "implementations" && $type !== "interfaces")
                continue;
            // skip e.g. bundle.json
            $pkgIdStr = $type . "/" . $pkgDir;
            $pkg = @$packages[$pkgIdStr] ?? null;
            if ($pkg === null) {
                $pkgDirParts = preg_split("/-/", $pkgDir);
                $version = preg_replace("/^v/", "", array_pop($pkgDirParts));
                $pkgType = $type === "implementations" ? PackageType::IMPLEMENTATION : PackageType::INTERFACE_;
                $pkgId = new PackageId($pkgType, implode("-", $pkgDirParts), $version);
                $pkg = new PackageContent($pkgId, Array(), true);
                $packages[$pkgIdStr] = $pkg;
            }
            $pkg->files[implode("/", $pathParts)] = OneFile::readText($this->packagesDir . "/" . $fn);
        }
        return new PackageBundle(array_values($packages));
    }
}

File: CSharp/StdLib/PackagesFolderSource.cs
using StdLib;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace StdLib
{
    public class PackagesFolderSource : PackageSource {
        public string packagesDir;
        
        public PackagesFolderSource(string packagesDir = "packages")
        {
            this.packagesDir = packagesDir;
        }
        
        public Task<PackageBundle> getPackageBundle(PackageId[] ids, bool cachedOnly)
        {
            throw new Error("Method not implemented.");
        }
        
        public async Task<PackageBundle> getAllCached()
        {
            var packages = new Dictionary<string, PackageContent> {};
            var allFiles = OneFile.listFiles(this.packagesDir, true);
            foreach (var fn in allFiles) {
                if (fn == "bundle.json")
                    continue;
                // TODO: hack
                var pathParts = fn.split(new RegExp("/")).ToList();
                // [0]=implementations/interfaces, [1]=package-name, [2:]=path
                var type = pathParts.shift();
                var pkgDir = pathParts.shift();
                if (type != "implementations" && type != "interfaces")
                    continue;
                // skip e.g. bundle.json
                var pkgIdStr = $"{type}/{pkgDir}";
                var pkg = packages.get(pkgIdStr);
                if (pkg == null) {
                    var pkgDirParts = pkgDir.split(new RegExp("-")).ToList();
                    var version = pkgDirParts.pop().replace(new RegExp("^v"), "");
                    var pkgType = type == "implementations" ? PackageType.Implementation : PackageType.Interface;
                    var pkgId = new PackageId(pkgType, pkgDirParts.join("-"), version);
                    pkg = new PackageContent(pkgId, new Dictionary<string, string> {}, true);
                    packages.set(pkgIdStr, pkg);
                }
                pkg.files.set(pathParts.join("/"), OneFile.readText($"{this.packagesDir}/{fn}"));
            }
            return new PackageBundle(Object.values(packages));
        }
    }
}